### DISCUSSION TITLE
Shadowing of variables

### DISCUSSION THREAD

**Discussion by MerlinEgalite** [morpho-org/morpho-optimizers#1582] [General]
Kickstarting a discussion for a general solution for the shadowing of variables.

Here are a few options already mentioned:
- Do nothing
- Add `Mem` to the end of a var.
- Adding an "_" to the beginning or the end of a var.

**Comment by Rubilmax** [comment]
@QGarchery also suggested (https://github.com/morpho-dao/morpho-v1/pull/1545#discussion_r1044498540):
- append `cached` suffix

I would suggest:
- `$` prefix for local, shadowing variables

**Reply by MerlinEgalite** [reply]
The prefix is used `$` is used by the Mangrove's team to represent address instead of an Interface typed variable.
Example:
```solidity
ERC20 public token;
address public $token;
```

@MathisGD was raising a small but a bit annoying issue with that which is that you cannot select the variable by just clicking on it.

**Reply by Rubilmax** [reply]
But Mangrove $\neq$ Morpho ; my point being we cannot also expect our code to conform to other teams' code conventions!
We didn't adopt this convention because we don't consider it necessary, so the syntax is free

Whereas prefixing `_` is already a syntax of ours, used for a different meaning 😝

**Reply by MerlinEgalite** [reply]
The issue raised is still true

**Reply by MathisGD** [reply]
I don't like the $ at all, it breaks my IDE assumptions

**Reply by MathisGD** [reply]
Thinking about it, adding something to all variables seems heavy for such a little issue (little in the sense that some people don't think that it is an issue).

**Reply by pakim249CAL** [reply]
I believe that this is actually a symptom of our lack of a strong naming convention that is free of any conflicts. A naming convention does not have to adhere to any other conventions, and should be suited to Morpho's specific needs. I am against doing nothing for V3, but I am ambivalent about any system we choose as long as the naming conventions are completely clear.

This link expresses the limits to our naming conventions pretty concisely.

https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityLexer.Identifier

Mangrove I believe is doing the right thing in making use of $. It is the only other available non-letter prefix available besides _. Shadowing is an issue that repeatedly comes up, so I'd like to see this issue solved so that discussions over variable naming is reduced, and this seems to be one of the only tools at our disposal.

**Reply by QGarchery** [reply]
> I don't like the $ at all, it breaks my IDE assumptions

The IDE makes wrong assumptions in this case, as the $ sign is in the regexp defining identifiers. This problem could probably be solved by raising an issue on the IDE side. That being said I think we both use vscode and I agree that this is annoying :/

**Reply by QGarchery** [reply]
I like prefixing with _ as I think the occurrences of needing both a local variable and input variable of the same name are pretty rare. It would also make our code look less like a christmas tree, with different weird symbols hooked all over the place :D

**Reply by pakim249CAL** [reply]
I don't think that case is very rare in our case. For example, take the entryPositionsManager set in the Morpho storage contracts.

This storage variable is currently named entryPositionManager. If it was made internal, this would be _entryPositionsManager. This would clash with the input variable in a setter, as the natural way to create a setter would be:

setEntryPositionsManager(address _entryPositionsManager)

So in this case, we can either use something like newEntryPositionsManager or _newEntryPositionsManager. Of those two options, I think I would prefer the former.

**Reply by QGarchery** [reply]
One of the reason input parameters are prefixed with _ is so that they don't clash with storage variables. So I don't think we want any storage variable (internal or not) to be prefixed with _
What do you think @morpho-dao/solidity @MathisGD ?

**Reply by pakim249CAL** [reply]
It can be known that input parameters if we do not add prefixes to these, would not clash with storage variables if ALL storage vars and internal functions are prefixed. So I am proposing actually that we abandon our practice of prefixing input variables.

**Comment by QGarchery** [comment]
The only problem I see with shadowing variables is that state variables may not be changed (because changes will apply to the shadowing variable), whereas the developer might expect changes . If the shadowing variable is not of the same type then I think it is fine to shadow, because the compiler will not accept wrong code. Do we agree that it is ok to shadow a state variable name with another variable of a different type ?

**Reply by MerlinEgalite** [reply]
I'm not worried about the compiler as he's doing a better job than humans. I'm more worried about humans having a harder time reviewing the code

**Reply by Rubilmax** [reply]
I'm ok to constrain the problem to local variables shadowing storage variables of the same type ; and that it's in this case not an issue for a reviewer

**Reply by MathisGD** [reply]
>humans having a harder time reviewing the code

Note that introducing new variables names for the "same thing" can also make reviewer's work harder

**Reply by QGarchery** [reply]
> I'm not worried about the compiler as he's doing a better job than humans. I'm more worried about humans having a harder time reviewing the code

I'm suggesting to rely more on the compiler, so do that mean you agree with my suggestion ?

**Reply by MerlinEgalite** [reply]
I think that we should first avoid declaring 2 variables with the same name even if their type differs.

**Reply by QGarchery** [reply]
Ok, then I guess I don't understand the link between this discussion and the comparison you make between humans and compilers :D 

I think it would be nice to have a second opinion on this from another reviewer

**Reply by MerlinEgalite** [reply]
I'm saying that the issue with shadowing is not because of the compiler but because it's harder for the human to read and it can be misleading too.

**Reply by QGarchery** [reply]
Personally I like to have short names, and if we are backed by the compiler in doing so (because it won't accept using the shadowing variable with the wrong type), it seems like a good idea to allow to keep the same name
